File: Resources/source/ScriptStore.lua
import ('System', 'System.IO')

local ScriptStore = {}

ScriptStore.mt = {__call = function(self, ...)
													return self.new (...)
											end
								  }
setmetatable (ScriptStore,ScriptStore.mt)

function ScriptStore.new (basePath, source)
	
	ScriptStore.m = {}; -- create a table members to store Lua fields
	
	if (String.IsNullOrEmpty (basePath)) then
		error ("Base path must be not empty");
	end
	
	if (String.IsNullOrEmpty (source)) then
		error ("Source path must be not empty");
	end

	if (not Directory.Exists (basePath)) then
			Directory.CreateDirectory (basePath);
	end

	ScriptStore.m.path = basePath;
	ScriptStore.m.sourcePath = source;
	
	return ScriptStore;
end

function ScriptStore:GetScripts()
	return Directory.GetFiles (self.m.path);
end

function ScriptStore:GetSources()
	return Directory.GetFiles (self.m.sourcePath);
end

function ScriptStore:SaveSourceContent (name, content)
			local filePath = Path.Combine (sourcePath, name);
			File.WriteAllText (filePath, content);
end

function ScriptStore:SaveScriptContent (name, content)

			local filePath = Path.Combine (self.m.path, name);
			File.WriteAllText (filePath, content);
end

function ScriptStore:Exists (name)
			local filePath = Path.Combine (self.m.path, name);
			return File.Exists (filePath);
end

function ScriptStore:RemoveFile (name)
			local filePath = Path.Combine (self.m.path, name);
			File.Delete (filePath);
end

function ScriptStore:RenameFile (oldName, newName)
			local path = self.m.path;
			local oldPath = Path.Combine (path, oldName);
			local newPath = Path.Combine (path, newName);
			if (File.Exists (newPath)) then
				File.Delete (newPath);
			end
			File.Move (oldPath, newPath);
end

function ScriptStore:GetScriptContent (name)
			local filePath = Path.Combine (self.m.path, name);
			return File.ReadAllText (filePath);
end

function ScriptStore:GetSourceContent (name)
			local filePath = Path.Combine (sourcePath, name);
			return File.ReadAllText (filePath);
end

return ScriptStore;

File: NLuaBox/ScriptStore.cs
using System;
using System.IO;
using System.Linq;

namespace NLuaBox
{
    public class ScriptStore
    {
        string path;

        public ScriptStore(string basePath)
        {
            if (string.IsNullOrEmpty(basePath))
                throw new ArgumentException(nameof(basePath));

            if (!Directory.Exists(basePath))
                Directory.CreateDirectory(basePath);

            path = basePath;
        }

        public string[] Scripts => Directory.GetFiles(path);

        public string GetFilePath(string name) => Path.Combine(path, name);

        public void WriteContent (string name, string content)
        {
            File.WriteAllText(GetFilePath(name), content);
        }

        public string ReadContent(string name)
        {
            return File.ReadAllText(GetFilePath(name));
        }

        public string ReadFirstLine(string name)
        {
            string filePath = GetFilePath(name);
            string line1 = File.ReadLines(filePath).FirstOrDefault();
            if (string.IsNullOrEmpty(line1))
                return " -- " + name;
            return line1;
        }

        public bool Exists(string name)
        {
            return File.Exists(GetFilePath(name));
        }

        public void RemoveFile(string name)
        {
            File.Delete(GetFilePath(name));
        }

        public void RenameFile(string oldName, string newName, string existingFileName)
        {
            string oldPath = GetFilePath(oldName);
            string newPath = GetFilePath(newName);
            string existingPath = GetFilePath(existingFileName);

            // Case renaming Foo -> foo, maybe is a bug on File.Move
            if (newName.Equals(oldName, StringComparison.OrdinalIgnoreCase))
            {
                string tempOld = Path.GetTempFileName();
                if (File.Exists(tempOld))
                    File.Delete(tempOld);
                File.Move(oldPath, tempOld);
                File.Move(tempOld, newPath);
                return;
            }

            if (File.Exists(existingPath))
                File.Delete(existingPath);

            File.Move(oldPath, newPath);
        }

        public static string FixName(string name)
        {
            name = name.Replace('?', '_');
            name = name.Replace(':', '_');
            name = name.Replace('/', '_');
            name = name.Replace('\\', '_');
            name = name.Replace('>', '_');
            name = name.Replace('<', '_');
            name = name.Replace('|', '_');

            if (name.EndsWith(".lua", StringComparison.OrdinalIgnoreCase))
                return name;
            return name + ".lua";
        }
    }
}
